src/components/QuranReader/TafsirView/TafsirBody.tsx

```diff
@@ -0,0 +1,211 @@
+import { useSelector, shallowEqual, useDispatch } from 'react-redux';
+import useSWR from 'swr/immutable';
+
+import SurahAndAyahLanguageSelection from './SurahAndAyahSelection';
+import TafsirGroupMessage from './TafsirGroupMessage';
+import TafsirAndLanguageSelection from './TafsirSelection';
+import TafsirSkeleton from './TafsirSkeleton';
+import styles from './TafsirView.module.scss';
+
+import { fetcher } from 'src/api';
+import DataFetcher from 'src/components/DataFetcher';
+import Separator from 'src/components/dls/Separator/Separator';
+import VerseText from 'src/components/Verse/VerseText';
+import { selectQuranReaderStyles } from 'src/redux/slices/QuranReader/styles';
+import { selectSelectedTafsirs, setSelectedTafsirs } from 'src/redux/slices/QuranReader/tafsirs';
+import QuranReaderStyles from 'src/redux/types/QuranReaderStyles';
+import { getDefaultWordFields } from 'src/utils/api';
+import { makeTafsirContentUrl, makeTafsirsUrl } from 'src/utils/apiPaths';
+import { areArraysEqual } from 'src/utils/array';
+import { fakeNavigate, getVerseTafsirNavigationUrl } from 'src/utils/navigation';
+import { getFirstAndLastFirstKeys, getVerseWords, makeVerseKey } from 'src/utils/verse';
+import { TafsirsResponse } from 'types/ApiResponses';
+
+type TafsirBodyProps = {
+  initialChapterId: string;
+  initialVerseNumber: string;
+  initialTafsirData?: TafsirsResponse;
+  initialTafsirId?: number;
+};
+
+const TafsirBody = ({
+  initialChapterId,
+  initialVerseNumber,
+  initialTafsirData,
+  initialTafsirId,
+}: TafsirBodyProps) => {
+  const dispatch = useDispatch();
+  const quranReaderStyles = useSelector(selectQuranReaderStyles, shallowEqual) as QuranReaderStyles;
+  const { lang } = useTranslation();
+  const userPreferredTafsirIds = useSelector(selectSelectedTafsirs, areArraysEqual);
+
+  const [selectedChapterId, setSelectedChapterId] = useState(initialChapterId);
+  const [selectedVerseNumber, setSelectedVerseNumber] = useState(initialVerseNumber);
+  const [selectedLanguage, setSelectedLanguage] = useState(lang);
+  const selectedVerseKey = makeVerseKey(Number(selectedChapterId), Number(selectedVerseNumber));
+  const [selectedTafsirId, setSelectedTafsirId] = useState(userPreferredTafsirIds?.[0]);
+
+  // if user opened tafsirBody via a url, we will have initialTafsirId
+  // we need to set this `initialTafsirId` as a selectedTafsirId
+  // we did not use `useState(initialTafsirId)` because `useRouter`'s query string is undefined on first render
+  useEffect(() => {
+    if (initialTafsirId) {
+      setSelectedTafsirId(initialTafsirId);
+    }
+  }, [initialTafsirId]);
+
+  const onTafsirSelected = useCallback(
+    (id: number) => {
+      setSelectedTafsirId(id);
+      fakeNavigate(
+        getVerseTafsirNavigationUrl(
+          Number(selectedChapterId),
+          Number(selectedVerseNumber),
+          id.toString(),
+        ),
+      );
+      dispatch(
+        setSelectedTafsirs({
```

Comment:
should this be an array?

Reply:
yeah, this is the one i mentioned yesterday about keeping the redux as array